R/utils.R

```diff
@@ -106,41 +106,70 @@ format_cmd_args <- function(.args, .collapse = FALSE) {
   }
 }
 
-#' Builds list of unique parameter sets for multiple model bbi calls
-#' @importFrom purrr map_lgl
-#' @param .mods a list containing only `bbi_{.model_type}_model` objects
-#' @param .bbi_args A named list specifying arguments to pass to babylon. This will over-ride any shared arguments from the model objects.
+#' Group models for bbi submission
+#'
+#' Multiple models can be submitted in a single bbi call if those models share a
+#' working directory and a set of CLI arguments.
+#'
+#' @inheritParams submit_nonmem_models
+#'
+#' @return A list whose elements correspond to groups of models that can be
+#'   submitted in a single bbi call. Each element is itself a list with elements
+#'
+#'   * `bbi_args`: a set of CLI arguments
+#'
+#'   * `models`: a list of the elements of `.mods` in the group
+#'
 #' @keywords internal
 build_bbi_param_list <- function(.mods, .bbi_args = NULL) {
 
   # check that everything in list is a model object
   check_model_object_list(.mods)
 
-  # build list of unique arg sets
-  param_list <- list()
-  for (.mod in .mods) {
-    # extract babylon args vector and working directory, then get md5 hash
-    args_vec <- parse_args_list(.bbi_args, .mod[[YAML_BBI_ARGS]]) %>% check_bbi_args() %>% sort()
-    model_dir <- .mod[[WORKING_DIR]]
-    arg_md5 <- c(args_vec, model_dir) %>% digest(algo = "md5")
-
-    if (is.null(param_list[[arg_md5]])) {
-      # create list for this arg set
-      param_list[[arg_md5]] <- list()
-
-      # add args vector and working directory
-      param_list[[arg_md5]][[YAML_BBI_ARGS]] <- args_vec
-      param_list[[arg_md5]][[WORKING_DIR]] <- model_dir
-    }
-
-    # append model path
-    param_list[[arg_md5]][[YAML_MOD_PATH]] <- c(param_list[[arg_md5]][[YAML_MOD_PATH]], .mod[[YAML_MOD_PATH]])
-  }
+  # character vector whose elements correspond to `.mods` and are strings
+  # representing the CLI args to bbi for the corresponding model
+  all_mod_args <-
+    .mods %>%
+    # TODO: consider whether this should be abstracted
+    purrr::map(YAML_BBI_ARGS) %>%
```

Comment:
I don't understand why these are all made as a their own `map` calls? Why not this?
```
map(.mods, function(.m) {
  .m[[YAML_BBI_ARGS]] %>%
  ...
})
```

The way it is strikes me as hard to read, but also doesn't this mean it will iterate over this list 4x instead of once?

Reply:
The goal of writing it with multiple `map()` calls is actually to improve readability and maintainability. This code can be run line by line, and after each line, the resulting list inspected. Each line is simple/does one thing. In addition, because all the `map()` calls are at the same level as the `.mods` object, we are guaranteed that after each transformation, the resulting list will have the same length as `.mods`. Thus, we can easily see the incremental transformations to `.mods`, which means it is easier to convince ourselves that those transformations are correct.

The pattern you describe means that the code cannot be run line by line interactively, because `.m` exists only within the scope of the anonymous function. For similar reasons, the result of each transformation cannot be inspected in turn.

So, yes, the way it is written results in more `map()` calls, but each call does something simple. Writing it with a single `map()` call and a more complicated anonymous function would also work, but is harder to run interactively, hence possibly harder to read and maintain.

Reply:
ok I'll buy that. If we were iterating over a large list this would be a concern, but it's negligible here. Thanks.